### DISCUSSION TITLE
Apply Clause

### DISCUSSION THREAD

**Discussion by ViliamVadocz** [AutoMapper/AutoMapper.Extensions.OData#107] [General]
I am going to be implementing the apply clause described in the [OData Specification](http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs02/odata-data-aggregation-ext-v4.0-cs02.html#_Toc435016567).

Related issue: #93 

I have some questions and will probably have more as I read more of the code.

1. Can I add to the [test data](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/master/AutoMapper.OData.EFCore.Tests/Data/DatabaseInitializer.cs)? I expect that the current data there will not have enough features to test aggregate operations.

2. Is there documentation for the internals of this code, preferably some kind of flowchart, that explains the flow of the code from query string to output collection?

3. Has any work been done on the apply clause that I should be aware of before starting?

4. Do you have any special contribution requirements? Is there a formatting specification I should follow?

**Comment by BlaiseD** [comment]
1. Yes - please include the [EF6 tests](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/master/AutoMapper.OData.EF6.Tests/Data/DatabaseInitializer.cs).
2. No.
3. Not in this repo.
4. Nope - just submit the PR and we'll go from there.

**Comment by ViliamVadocz** [comment]
Is there a reason for the code repetition in `AutoMapper.AspNet.OData.QueryableExtensions` methods `Get`, `GetAsync`, `GetQuery`, `GetQueryAsync`?

I assume this is where I will need to make changes. Currently the code only considers `options.Filter`, but for apply I will need to look at `options.Apply`. Is my understanding correct?

**Reply by BlaiseD** [reply]
I don't see repetition in those methods.  Either way your changes should only address new logical changes.

Almost.  You'll find multiple properties for `ODataQueryOptions` already being used:
![image](https://user-images.githubusercontent.com/7595209/129361579-b0130a83-4def-44e2-bbca-6cd0737f9d39.png)

Where the changes go depends on your research.  I'm guessing they would go in the Get* methods with the `ODataQueryOptions<TModel> options` in the signature applying the new aggregation code to the current results.

**Reply by BlaiseD** [reply]
`SelectExpand`'s being used not Apply.

**Comment by ViliamVadocz** [comment]
It looks like `Microsoft.AspNet.OData` only supports a subset of what is defined in the OData specification. 
For example, the specification defines [`$apply=identity`](http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs02/odata-data-aggregation-ext-v4.0-cs02.html#_Toc435016580), but there isn't an identity [`TransformationNode`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.odata.uriparser.aggregation.transformationnode?view=odata-core-7.0) defined.

I will only add tests for what is supported.

**Reply by BlaiseD** [reply]
Not sure what you mean:
![image](https://user-images.githubusercontent.com/7595209/129362863-b0275a33-69e3-4255-a9da-8b88ab06af6d.png)

![image](https://user-images.githubusercontent.com/7595209/129362927-669b6ddc-ec22-4b05-8ffe-d1ae2e2e8bd9.png)

Adding logic for only what's supported sounds about right.

**Comment by ViliamVadocz** [comment]
I am still scratching my head trying to figure this out.
Why do you apply the operations like `$filter` using expressions instead of just using [`ODataQueryOptions.ApplyTo`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnet.odata.query.odataqueryoptions.applyto?view=odata-aspnetcore-7.0#Microsoft_AspNet_OData_Query_ODataQueryOptions_ApplyTo_System_Linq_IQueryable_)? Is something not supported with `ApplyTo` or is there some reason related to automatic mapping?

**Reply by ViliamVadocz** [reply]
I am currently looking at [some tests](https://github.com/OData/WebApi/blob/master/test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Aggregation/AggregationTests.cs) for usage examples, but I can't find the implementation. I don't understand how I apply the parsed request. With `$filter` you apply it onto an empty queryable and then turn it into an expression. With `$expand` you do something seemingly completely different.

**Reply by BlaiseD** [reply]
Ideally with filter we'd want to do the following:
```c#
        public static Expression<Func<T, bool>> ToFilterExpression<T>(this FilterQueryOption filterOption, HandleNullPropagationOption handleNullPropagation = HandleNullPropagationOption.Default)
        {
            if (filterOption == null)
                return null;

            return (Expression<Func<T, bool>>)filterOption.FilterClause.GetFilterExpression(typeof(T));

            //IQueryable queryable = Enumerable.Empty<T>().AsQueryable();
            //queryable = filterOption.ApplyTo(queryable, new ODataQuerySettings() { HandleNullPropagation = handleNullPropagation });
            //MethodCallExpression whereMethodCallExpression = (MethodCallExpression)queryable.Expression;
            //return (Expression<Func<T, bool>>)(whereMethodCallExpression.Arguments[1].Unquote() as LambdaExpression);
        }
```

so that we're creating all our own expressions.

I you run this you'll find a couple of cases unhandled.    We use `GetFilterExpression` for filtering child collections.

If all the test cases work using `ODataQueryOptions.ApplyTo` combined with `ProjectTo` then maybe we're doing too much work here.

On the other hand historically with `ProjectTo`, applying the filter [before hand ](https://github.com/AutoMapper/AutoMapper/issues/2611)was recommended.  You might might find some failing cases otherwise.

**Reply by BlaiseD** [reply]
With regard to applying the parsed request, it depends on what you come up with.  I would think something like we're doing for [OrderBy](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/c196308e7c15c1ed82653c60e396fce2df4fa076/AutoMapper.AspNetCore.OData.EFCore/LinqExtensions.cs#L85) i.e. create the `GroupBy` or aggregate expressions based on the parsed data.

**Reply by ViliamVadocz** [reply]
`options.ApplyTo(mapper.ProjectTo<TModel>(query))` doesn't immediately error, but later on gives
```
Processing of the LINQ expression '(GroupByShaperExpression:
KeySelector: new NoGroupByWrapper(), 
ElementSelector:new OpsTenant{ }
)' by 'InMemoryProjectionBindingExpressionVisitor' failed. This may indicate either a bug or a limitation in EF Core. See https://go.microsoft.com/fwlink/?linkid=2101433 for more detailed information.
````

This is probably related to the issue you linked.

(trying `"/opstenant?$apply=aggregate($count as TenantCount)"`)

**Reply by BlaiseD** [reply]
Can't say for sure - that looks like an expression created by/executed in `ApplyTo`. `InMemoryProjectionBindingExpressionVisitor` looks it's trying to aggregate on the client and failing.  (the link also points to Client vs. Server Evaluation)

In my opinion `ApplyTo` + `ProjectTo` + `EF` makes fixes more difficult. 

Building most of the expressions ourselves is more work but makes adjustment easier i.e. you'd have control over what to do about that expression.

**Reply by ViliamVadocz** [reply]
I have the issue that I don't understand how to build the correct expression and how to use it. I could do something similar to `$filter` where I  use `ApplyClause.ApplyTo` on an empty queryable and turn that into an expression, but I don't see how to then connect it to the rest of the operations. From what I read, `$apply` is should be evaluated first and then other things operate on the result of that: http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs02/odata-data-aggregation-ext-v4.0-cs02.html#_Toc435016590
Seems to me like [some intermediate type](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnet.odata.query.expressions.dynamictypewrapper?view=odata-aspnetcore-7.0) will be required, but I probably just don't understand it well enough yet.

**Reply by BlaiseD** [reply]
The library references an [expression helper](https://github.com/BlaiseD/LogicBuilder.DataComponents/blob/601d1366ca7fe32a0eebe3f0102852e5c586fa83/LogicBuilder.Expressions.Utils/LinqHelpers.cs#L444) which may be helpful.

When the expression type changes you use type of the current expression to build the next one as illustrated in the linked `GroupByCall` example.  I can't say what the solution is without digging in. 

It's not always solvable in one day.  Consider starting with the parameter [expression here](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/c196308e7c15c1ed82653c60e396fce2df4fa076/AutoMapper.AspNetCore.OData.EFCore/LinqExtensions.cs#L73).  Add a `GroupBy` to it.

Then map the expression [like here](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/c196308e7c15c1ed82653c60e396fce2df4fa076/AutoMapper.AspNetCore.OData.EFCore/QueryableExtensions.cs#L111) and run it.

From there figure out what is or isn't possible using `ProjectTo` with `GroupBy` and the aggregate functions - also whether using the existing functions with the new code will be possible.  Once you get a couple of snippets working then a solution could become apparent.
